### PR TITLE
minor optimization for userobject based crystal plasticity

### LINKED ISSUE #6706: Minor optimize userobject based crystal plasticity
### Description of the enhancement or error report
1. UO based CP does not work for material properties on the boundaries. Add a check to skip if it is on boundary. 
2. Minor changes made to CrystalPlasticityStateVariable userobject. it is now more general and have better data reading method.
3. some clean up for FiniteStrainUOBasedCrystalPlasticity  

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)
Minor local changes.

### PR DESCRIPTION
**Pull request by jiangwen84**
Some tests are modified and a new test is added. 

closes #6706

### CODE REVIEW

**Comment by dschwen** [comment]
Documentation for this system is here:
http://mooseframework.org/wiki/PhysicsModules/TensorMechanics/CrystalPlasticity/

**Comment by permcody** [comment]
@dschwen - please review

File: modules/tensor_mechanics/include/userobjects/CrystalPlasticityStateVariable.h [hunk 1]
```diff
@@ -35,12 +37,16 @@ class CrystalPlasticityStateVariable : public CrystalPlasticityUOBase
   const MaterialProperty<std::vector<Real> > &  _mat_prop_state_var;
   const MaterialProperty<std::vector<Real> > &  _mat_prop_state_var_old;
 
-  /// File should contain initial values of the slip system resistances.
-  std::string _slip_sys_res_prop_file_name;
+  /// File should contain initial values of the state variable.
+  std::string _state_variable_file_name;
```
Comment (dschwen) [inline modules/tensor_mechanics/include/userobjects/CrystalPlasticityStateVariable.h]: Would make this of type `FileName`

File: modules/tensor_mechanics/include/userobjects/CrystalPlasticityStateVariable.h [hunk 2]
```diff
@@ -35,12 +37,16 @@ class CrystalPlasticityStateVariable : public CrystalPlasticityUOBase
...
 
   /// Read from options for initial values of internal variables
   MooseEnum _intvar_read_type;
 
+  std::vector<unsigned int> _groups;
```
Comment (dschwen) [inline modules/tensor_mechanics/include/userobjects/CrystalPlasticityStateVariable.h]: can you add a doc comment here?

File: modules/tensor_mechanics/src/userobjects/CrystalPlasticityStateVariable.C [hunk 3]
```diff
@@ -42,24 +46,58 @@ CrystalPlasticityStateVariable::CrystalPlasticityStateVariable(const InputParame
...
+   if (is > ie)
+      mooseError("CrystalPlasticityStateVariable: Start index is = " << is << " should be greater than end index ie = " << ie << " in state variable read");
+
+    for (unsigned int j = is; j <= ie; ++j)
+      val[j] = _group_values[i];
+  }
+
+  for (unsigned int i = 0; i < _variable_size; ++i)
+    if (val[i] <= 0.0)
+      mooseError("CrystalPlasticityStateVariable: Value of state variables " << i  << " non positive");
```
Comment (dschwen) [inline modules/tensor_mechanics/src/userobjects/CrystalPlasticityStateVariable.C]: why do you only check this in this function and not in the read from file function? Would it make sense to move the check to right after the switch block in the constructor?

File: modules/tensor_mechanics/src/userobjects/CrystalPlasticityStateVariable.C [hunk 4]
```diff
@@ -42,24 +46,58 @@ CrystalPlasticityStateVariable::CrystalPlasticityStateVariable(const InputParame
 void
 CrystalPlasticityStateVariable::initSlipSysProps(std::vector<Real> & val) const
 {
-  readFileInitSlipSysRes(val);
+  switch (_intvar_read_type)
```
Comment (dschwen) [inline modules/tensor_mechanics/src/userobjects/CrystalPlasticityStateVariable.C]: This is just my preference, but I would put in two `case` statements (one for each valid option) and have the `default` error out. This makes sure the won't be any surprises when someone modifies the moose enum above and forgets the switch, or if there is an illegal memory access that overwrites the variable etc.

**Comment by dschwen** [comment]
Apart from these minor comments this looks really good!

**Comment by jiangwen84** [comment]
@dschwen Thanks for your comments. I will change them accordingly. -:)

**Comment by jiangwen84** [comment]
Address the comments from @dschwen 

Also fixed the compiler warning in my code #6720 @permcody

**Pull request merged**